ConcurrentModificationException while processing REST request

I discovered this in the Logs of my openHAB Server:

```
2021-10-17 09:21:16.006 [ERROR] [internal.JSONResponseExceptionMapper] - Unexpected exception occurred while processing REST request.
java.util.ConcurrentModificationException: null
        at java.util.LinkedHashMap$LinkedKeySet.forEach(LinkedHashMap.java:561) ~[?:?]
        at org.openhab.core.io.rest.auth.internal.ExpiringUserSecurityContextCache.get(ExpiringUserSecurityContextCache.java:53) ~[?:?]
        at org.openhab.core.io.rest.auth.internal.AuthFilter.authenticateBasicAuth(AuthFilter.java:189) ~[?:?]
        at org.openhab.core.io.rest.auth.internal.AuthFilter.filter(AuthFilter.java:252) ~[?:?]
        at org.apache.cxf.jaxrs.utils.JAXRSUtils.runContainerRequestFilters(JAXRSUtils.java:1745) ~[bundleFile:3.4.3]
        at org.apache.cxf.jaxrs.interceptor.JAXRSInInterceptor.processRequest(JAXRSInInterceptor.java:110) ~[bundleFile:3.4.3]
        at org.apache.cxf.jaxrs.interceptor.JAXRSInInterceptor.handleMessage(JAXRSInInterceptor.java:79) ~[bundleFile:3.4.3]
        at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:308) ~[bundleFile:3.4.3]
        at org.apache.cxf.transport.ChainInitiationObserver.onMessage(ChainInitiationObserver.java:121) ~[bundleFile:3.4.3]
        at org.apache.cxf.transport.http.AbstractHTTPDestination.invoke(AbstractHTTPDestination.java:265) ~[bundleFile:3.4.3]
        at org.apache.cxf.transport.servlet.ServletController.invokeDestination(ServletController.java:234) ~[bundleFile:3.4.3]
        at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:208) ~[bundleFile:3.4.3]
        at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:160) ~[bundleFile:3.4.3]
        at org.apache.cxf.transport.servlet.CXFNonSpringServlet.invoke(CXFNonSpringServlet.java:225) ~[bundleFile:3.4.3]
        at org.apache.cxf.transport.servlet.AbstractHTTPServlet.handleRequest(AbstractHTTPServlet.java:298) ~[bundleFile:3.4.3]
        at org.apache.cxf.transport.servlet.AbstractHTTPServlet.doGet(AbstractHTTPServlet.java:222) ~[bundleFile:3.4.3]
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:687) ~[bundleFile:3.1.0]
        at org.apache.cxf.transport.servlet.AbstractHTTPServlet.service(AbstractHTTPServlet.java:273) ~[bundleFile:3.4.3]
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:799) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:550) ~[bundleFile:9.4.43.v20210629]
        at org.ops4j.pax.web.service.jetty.internal.HttpServiceServletHandler.doHandle(HttpServiceServletHandler.java:71) ~[bundleFile:?]
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:602) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1434) ~[bundleFile:9.4.43.v20210629]
        at org.ops4j.pax.web.service.jetty.internal.HttpServiceContext.doHandle(HttpServiceContext.java:294) ~[bundleFile:?]
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:501) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1349) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) ~[bundleFile:9.4.43.v20210629]
        at org.ops4j.pax.web.service.jetty.internal.JettyServerHandlerCollection.handle(JettyServerHandlerCollection.java:82) ~[bundleFile:?]
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.server.Server.handle(Server.java:516) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:388) ~[bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:633) [bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:380) [bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277) [bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311) [bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105) [bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104) [bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338) [bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315) [bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173) [bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131) [bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:386) [bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883) [bundleFile:9.4.43.v20210629]
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034) [bundleFile:9.4.43.v20210629]
        at java.lang.Thread.run(Thread.java:829) [?:?]
```

Unfortunately I don't know what caused this or how this could be reproduced.

--- a/bundles/org.openhab.core.io.rest.auth/src/main/java/org/openhab/core/io/rest/auth/internal/ExpiringUserSecurityContextCache.java
+++ b/bundles/org.openhab.core.io.rest.auth/src/main/java/org/openhab/core/io/rest/auth/internal/ExpiringUserSecurityContextCache.java
@@ -12,6 +12,7 @@
  */
 package org.openhab.core.io.rest.auth.internal;
 
+import java.util.HashSet;
 import java.util.LinkedHashMap;
 import java.util.Map;
 
@@ -28,8 +29,8 @@ import org.eclipse.jdt.annotation.Nullable;
  */
 @NonNullByDefault
 public class ExpiringUserSecurityContextCache {
-    private static final int MAX_SIZE = 10;
-    private static final int CLEANUP_FREQUENCY = 10;
+    static final int MAX_SIZE = 10;
+    static final int CLEANUP_FREQUENCY = 10;
 
     private final long keepPeriod;
     private final Map<String, Entry> entryMap;
@@ -50,7 +51,7 @@ public class ExpiringUserSecurityContextCache {
     synchronized @Nullable UserSecurityContext get(String key) {
         calls++;
         if (calls >= CLEANUP_FREQUENCY) {
-            entryMap.keySet().forEach(k -> getEntry(k));
+            new HashSet<>(entryMap.keySet()).forEach(k -> getEntry(k));
             calls = 0;
         }
         Entry entry = getEntry(key);

--- a/bundles/org.openhab.core.io.rest.auth/src/test/java/org/openhab/core/io/rest/auth/internal/ExpiringUserSecurityContextCacheTest.java
+++ b/bundles/org.openhab.core.io.rest.auth/src/test/java/org/openhab/core/io/rest/auth/internal/ExpiringUserSecurityContextCacheTest.java
@@ -1,0 +1,132 @@
+/**
+ * Copyright (c) 2010-2021 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.core.io.rest.auth.internal;
+
+import static org.hamcrest.CoreMatchers.*;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.openhab.core.io.rest.auth.internal.ExpiringUserSecurityContextCache.*;
+
+import java.time.Duration;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.junit.jupiter.api.Test;
+import org.openhab.core.auth.Authentication;
+import org.openhab.core.auth.GenericUser;
+
+/**
+ * Tests {@link ExpiringUserSecurityContextCache}.
+ *
+ * @author Wouter Born - Initial contribution
+ */
+@NonNullByDefault
+public class ExpiringUserSecurityContextCacheTest {
+
+    private static final Duration ONE_HOUR = Duration.ofHours(1);
+
+    private ExpiringUserSecurityContextCache createCache(Duration expirationDuration) {
+        return new ExpiringUserSecurityContextCache(expirationDuration.toMillis());
+    }
+
+    private Map<String, UserSecurityContext> createValues(int count) {
+        Map<String, UserSecurityContext> map = new LinkedHashMap<>();
+        for (int i = 0; i < count; i++) {
+            String userName = "user" + i;
+            UserSecurityContext userSecurityContext = new UserSecurityContext(new GenericUser(userName),
+                    new Authentication(userName), userName + " token");
+            map.put("key" + i, userSecurityContext);
+        }
+        return map;
+    }
+
+    private void addValues(ExpiringUserSecurityContextCache cache, Map<String, UserSecurityContext> values) {
+        values.entrySet().stream().forEach(entry -> cache.put(entry.getKey(), entry.getValue()));
+    }
+
+    private void assertValuesAreCached(Map<String, UserSecurityContext> values,
+            ExpiringUserSecurityContextCache cache) {
+        for (Entry<String, UserSecurityContext> entry : values.entrySet()) {
+            assertThat(cache.get(entry.getKey()), is(entry.getValue()));
+        }
+    }
+
+    private void assertValuesAreNotCached(Map<String, UserSecurityContext> values,
+            ExpiringUserSecurityContextCache cache) {
+        for (Entry<String, UserSecurityContext> entry : values.entrySet()) {
+            assertThat(cache.get(entry.getKey()), is(nullValue()));
+        }
+    }
+
+    @Test
+    public void cachedValuesAreReturned() {
+        ExpiringUserSecurityContextCache cache = createCache(ONE_HOUR);
+        Map<String, UserSecurityContext> values = createValues(MAX_SIZE);
+        addValues(cache, values);
+        assertValuesAreCached(values, cache);
+    }
+
+    @Test
+    public void nonCachedValuesAreNotReturned() {
+        ExpiringUserSecurityContextCache cache = createCache(ONE_HOUR);
+        Map<String, UserSecurityContext> values = createValues(MAX_SIZE);
+        assertValuesAreNotCached(values, cache);
+    }
+
+    @Test
+    public void clearedValuesAreNotReturned() {
+        ExpiringUserSecurityContextCache cache = createCache(ONE_HOUR);
+        Map<String, UserSecurityContext> values = createValues(MAX_SIZE);
+        addValues(cache, values);
+        cache.clear();
+        assertValuesAreNotCached(values, cache);
+    }
+
+    @Test
+    public void eldestEntriesAreRemovedWhenMaxSizeIsExceeded() {
+        ExpiringUserSecurityContextCache cache = createCache(ONE_HOUR);
+        int removed = 20;
+        Map<String, UserSecurityContext> values = createValues(MAX_SIZE + removed);
+        addValues(cache, values);
+
+        Map<String, UserSecurityContext> removedValues = new LinkedHashMap<>();
+        Map<String, UserSecurityContext> cachedValues = new LinkedHashMap<>();
+
+        int i = 0;
+        for (Entry<String, UserSecurityContext> entry : values.entrySet()) {
+            if (i < removed) {
+                removedValues.put(entry.getKey(), entry.getValue());
+            } else {
+                cachedValues.put(entry.getKey(), entry.getValue());
+            }
+            i++;
+        }
+
+        assertValuesAreNotCached(removedValues, cache);
+        assertValuesAreCached(cachedValues, cache);
+    }
+
+    @Test
+    public void expiredEntriesAreRemoved() {
+        ExpiringUserSecurityContextCache cache = createCache(Duration.ZERO);
+        Map<String, UserSecurityContext> values = createValues(MAX_SIZE);
+        addValues(cache, values);
+
+        for (int i = 0; i < CLEANUP_FREQUENCY; i++) {
+            cache.get("key");
+        }
+
+        assertValuesAreNotCached(values, cache);
+    }
+}

